extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/logical/LogicalStageBuilder.java

```diff
@@ -0,0 +1,273 @@
+        DruidFilter filter = (DruidFilter) stack.peekNode();
+        return makeFilterStage(filter);
+      }
+
+      if (stack.peekNode() instanceof DruidProject) {
+
+        DruidProject project = (DruidProject) stack.peekNode();
+        DruidFilter dummyFilter = new DruidFilter(
+            project.getCluster(), project.getTraitSet(), project,
+            project.getCluster().getRexBuilder().makeLiteral(true)
+        );
+        return makeFilterStage(dummyFilter).extendWith(stack);
+      }
+      return null;
+    }
+
+    private LogicalStage makeFilterStage(DruidFilter filter)
+    {
+      VirtualColumnRegistry virtualColumnRegistry = VirtualColumnRegistry.create(
+          signature,
+          plannerContext.getExpressionParser(),
+          plannerContext.getPlannerConfig().isForceExpressionVirtualColumns()
+      );
+
+      DimFilter dimFilter = DruidQuery.getDimFilter(
+          plannerContext,
+          signature, virtualColumnRegistry, filter
+      );
+
+      return new FilterStage(
+          this,
+          virtualColumnRegistry,
+          dimFilter
+      );
+    }
+  }
+
+  public FilterStage create(RootStage inputStage, DruidFilter filter)
+  {
+    VirtualColumnRegistry virtualColumnRegistry = VirtualColumnRegistry.create(
+        inputStage.signature,
+        plannerContext.getExpressionParser(),
+        plannerContext.getPlannerConfig().isForceExpressionVirtualColumns()
+    );
+    DimFilter dimFilter = DruidQuery.getDimFilter(plannerContext, inputStage.signature, virtualColumnRegistry, filter);
+    return new FilterStage(inputStage, virtualColumnRegistry, dimFilter);
+  }
+
+  class FilterStage extends RootStage
+  {
+    protected final VirtualColumnRegistry virtualColumnRegistry;
+    private DimFilter dimFilter;
+
+    public FilterStage(RootStage inputStage, VirtualColumnRegistry virtualColumnRegistry, DimFilter dimFilter)
+    {
+      super(inputStage, inputStage.signature);
+      this.virtualColumnRegistry = virtualColumnRegistry;
+      this.dimFilter = dimFilter;
+    }
+
+    public FilterStage(FilterStage root, VirtualColumnRegistry newVirtualColumnRegistry, RowSignature rowSignature)
+    {
+      super(root, rowSignature);
+      this.dimFilter = root.dimFilter;
+      this.virtualColumnRegistry = newVirtualColumnRegistry;
+    }
+
+    @Override
+    public StageDefinition buildCurrentStage(StageMaker stageMaker)
+    {
+      VirtualColumns output = virtualColumnRegistry.build(Collections.emptySet());
+      return stageMaker.makeScanStage(output, signature, inputSpecs, dimFilter);
+    }
+
+    @Override
+    public LogicalStage extendWith(DruidNodeStack stack)
+    {
+      if (stack.peekNode() instanceof DruidProject) {
+        DruidProject project = (DruidProject) stack.peekNode();
+        Projection preAggregation = Projection
+            .preAggregation(project, plannerContext, signature, virtualColumnRegistry);
+
+        return new ProjectStage(
```

Comment:
Is it ok that this is only considering the first node in the stack? Would other items in the stack be ignored or will they be handled somewhere else?

Reply:
yes - maybe the name of the parameter should be changed to `parents` or `parentStack` ?
and/or reorganize the `DruidNodeStack` class a bit - so that its not that confusing that it also provides access to further things